src/plugins/controls/public/services/kibana/options_list.ts

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { memoize } from 'lodash';
+import dateMath from '@elastic/datemath';
+import { buildEsQuery } from '@kbn/es-query';
+
+import { pluginServices } from '../../services';
+import { TimeRange } from '../../../../data/public';
+import { ControlsOptionsListService, OptionsListRequest } from '../options_list';
+import {
+  OptionsListRequestBody,
+  OptionsListResponse,
+} from '../../control_types/options_list/types';
+import { KibanaPluginServiceFactory } from '../../../../presentation_util/public';
+import { ControlsPluginStartDeps } from '../../types';
+
+const getRoundedTimeRange = (timeRange: TimeRange) => ({
+  from: dateMath.parse(timeRange.from)!.startOf('minute').toISOString(),
+  to: dateMath.parse(timeRange.to)!.endOf('minute').toISOString(),
+});
+
+const optionsListCacheResolver = (request: OptionsListRequest) => {
+  const {
+    query,
+    filters,
+    timeRange,
+    searchString,
+    selectedOptions,
+    field: { name: fieldName },
+    dataView: { title: dataViewTitle },
+  } = request;
+  return [
+    ...(timeRange ? JSON.stringify(getRoundedTimeRange(timeRange)) : []), // round timeRange to the minute to avoid cache misses
+    Math.floor(Date.now() / 1000 / 60), // Only cache results for a minute in case data changes in ES index
+    selectedOptions?.join(','),
+    JSON.stringify(filters),
+    JSON.stringify(query),
+    dataViewTitle,
+    searchString,
+    fieldName,
+  ].join('|');
+};
+
+const cachedOptionsListRequest = memoize(
+  async (request: OptionsListRequest, abortSignal: AbortSignal) => {
+    const { http } = pluginServices.getServices();
```

Comment:
I know there was some work done to setup some rudimentary dependecy injection between services.  So, you should be able to list the `http` service as a dependency of the options list service and it will get injected in. Might be nicer than the `getServices()` option.  

Example definition and usage: https://github.com/elastic/kibana/blob/main/x-pack/plugins/canvas/public/services/kibana/index.ts#L46

https://github.com/elastic/kibana/blob/main/x-pack/plugins/canvas/public/services/kibana/expressions.ts#L139

Reply:
Very cool! That made the service a lot cleaner, thanks for showing me this!